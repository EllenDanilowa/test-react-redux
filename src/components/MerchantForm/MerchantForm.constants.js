const EMAIL_REGEX = RegExp(/^(([^<>()[\].,;:\s@"]+(\.[^<>()[\].,;:\s@"]+)*)|(".+"))@(([^<>()[\].,;:\s@"]+\.)+[^<>()[\].,;:\s@"]{2,})$/i);

export const FIELDS = {
  FIRST_NAME: {
    name: 'firstname',
    label: 'First name',
    defaultValue: '',
    rules: {required: true, minLength: 2, maxLength: 50},
    errorMessage: 'Please enter valid first name'
  },
  LAST_NAME: {
    name: 'lastname',
    label: 'Last name',
    defaultValue: '',
    rules: {required: true, minLength: 2, maxLength: 50},
    errorMessage: 'Please enter valid last name'
  },
  EMAIL: {
    name: 'email',
    label: 'Email',
    defaultValue: '',
    rules: {required: true, pattern: EMAIL_REGEX},
    errorMessage: 'Please enter valid email'
  },
  PHONE: {
    name: 'phone',
    label: 'Phone number',
    defaultValue: '',
    rules: {required: true},
    errorMessage: 'Please enter valid phone number'
  },
  HAS_PREMIUM: {
    name: 'hasPremium',
    label: 'Premium',
    defaultValue: false
  },
  AVATAR: {
    name: 'avatar',
    label: 'Avatar'
  }
};
