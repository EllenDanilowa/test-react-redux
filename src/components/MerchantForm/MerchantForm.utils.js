export const createFormInitialState = (fields, initObj = {}) => {
  const state = {
    ...initObj,
    errors: {}
  };
  Object.values(fields).forEach((field) => {
    const {name, defaultValue} = field;

    state[name] = state[name] || defaultValue;
    state.errors[name] = '';
  });

  return state;
};

export const checkFormValidation = (errors) => {
  let valid = true;

  Object.values(errors).forEach((val) => val.length && (valid = false));

  return valid;
};
