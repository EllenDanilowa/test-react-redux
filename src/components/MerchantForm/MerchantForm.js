import React, {Component} from 'react';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';
import Input from '../Form/Input/Input';
import Button from '../Form/Button/Button';
import {Title} from './MerchantForm.styled';
import {
  DEFAULT_ERROR_MESSAGE,
  FIELDS
} from './MerchantForm.constants';

const createInitialState = (initObj = {}) => {
  const state = {errors: {}};
  Object.values(FIELDS).forEach((field) => {
    const {name, defaultValue} = field;

    state[name] = initObj[name] || defaultValue;
    state.errors[name] = '';
  });

  return state;
};

const validateForm = (errors) => {
  let valid = true;

  Object.values(errors).forEach((val) => (
    val.length && (valid = false)
  ));

  return valid;
};

class MerchantForm extends Component {
  constructor(props) {
    super(props);

    this.state = createInitialState(this.props.initValue);

    this.onInputChange = this.onInputChange.bind(this);
    // this.onAvatarChange = this.onAvatarChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  validateForm() {
    const errors = {};

    Object.values(FIELDS).forEach((field) => {
      const {rules = {}, name} = field;
      const value = this.state[field.name];

      if (rules.required) {
        errors[name] = value.length === 0;
      }

      if (rules.max) {
        errors[name] = value.length > rules.max;
      }

      if (rules.min) {
        errors[name] = value.length < rules.min;
      }

      if (rules.regexp) {
        errors[name] = !rules.regexp.test(value);
      }

      errors[name] = errors[name] ? (field.errorMessage || DEFAULT_ERROR_MESSAGE) : '';
    });

    this.setState({errors});

    return validateForm(errors);
  }

  onInputChange(event) {
    const {name, value} = event.target;
    const {errors} = this.state;

    errors[name] = '';
    this.setState({errors, [name]: value});
  }

  //
  // onAvatarChange(event) {
  //   this.setState({avatar: event.target.files[0]});
  // }

  onSubmit(event) {
    event.preventDefault();

    if (this.validateForm()) {
      this.props.submit(this.state);
    }
  }

  render() {
    return (
      <div>
        {this.props.title && (
          <Title>{this.props.title}</Title>
        )}
        <form onSubmit={this.onSubmit} noValidate>
          <Input name="firstname"
                 title="First name"
                 error={this.state.errors.firstname}
                 value={this.state.firstname}
                 onChange={this.onInputChange}/>
          <Input name="lastname"
                 title="Last name"
                 error={this.state.errors.lastname}
                 value={this.state.lastname}
                 onChange={this.onInputChange}/>
          <Input name="email"
                 title="Email"
                 error={this.state.errors.email}
                 value={this.state.email}
                 onChange={this.onInputChange}/>
          <Input name="phone"
                 type="tel"
                 title="Phone"
                 error={this.state.errors.phone}
                 value={this.state.phone}
                 onChange={this.onInputChange}/>
          <Input name="hasPremium"
                 type="checkbox"
                 title="Premium"
                 value={this.state.hasPremium}
                 onChange={this.onInputChange}/>

          {/*<div>*/}
          {/*  <label htmlFor="avatar">Avatar</label>*/}
          {/*  <input type="file" name="avatar" id="avatar" accept="image/*" ref={this.avatarRef}*/}
          {/*         onChange={this.onAvatarChange}/>*/}
          {/*</div>*/}

          <Button title={this.props.submitTitle}/>
        </form>
        <Link to="/">Cancel</Link>
      </div>
    );
  }
}

MerchantForm.propTypes = {
  initValue: PropTypes.shape({
    id: PropTypes.string,
    firstname: PropTypes.string.isRequired,
    lastname: PropTypes.string.isRequired,
    avatarUrl: PropTypes.string,
    email: PropTypes.string.isRequired,
    phone: PropTypes.string.isRequired,
    hasPremium: PropTypes.bool.isRequired,
    bids: PropTypes.array
  }),
  submit: PropTypes.func.isRequired,
  submitTitle: PropTypes.string.isRequired,
  title: PropTypes.string
};

export default MerchantForm;
