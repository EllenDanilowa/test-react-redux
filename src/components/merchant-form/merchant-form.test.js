import React from 'react';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router';
import MerchantForm from './merchant-form';
import {mount} from "enzyme";
import {getDefaultMerchant} from './merchant-form.utils';

describe('MerchantForm', () => {
  let element;
  let submit;
  let submitTitle;

  const createElement = (props, isSnapshot = true) => {
    const template = (<MemoryRouter><MerchantForm {...props}/></MemoryRouter>);

    return isSnapshot ?
      renderer.create(template).toJSON() :
      mount(template);
  };

  beforeEach(() => {
    submit = jest.fn();
    submitTitle = 'Submit';
  });

  it('renders component with defined merchant', () => {
    const merchant = {
      id: '1',
      firstname: 'name',
      lastname: 'lastname',
      avatarUrl: 'avatar/url',
      email: 'test@test.com',
      phone: '123456789',
      hasPremium: true,
      bids: []
    };
    element = createElement({submit, submitTitle, merchant});

    expect(element).toMatchSnapshot();
  });

  it('renders component by default', () => {
    element = createElement({submit, submitTitle});

    expect(element).toMatchSnapshot();
  });
});

describe('MerchantForm utils', () => {
  describe('getDefaultMerchant', () => {
    it('returns default state from settings', () => {
      const settings = {
        name: {name: 'name', defaultValue: 'test'},
        phone: {name: 'phone', defaultValue: '23456789'}
      };
      const expectedObj = {
        name: 'test',
        phone: '23456789'
      };

      expect(getDefaultMerchant(settings)).toEqual(expectedObj);
    });
  });
});
