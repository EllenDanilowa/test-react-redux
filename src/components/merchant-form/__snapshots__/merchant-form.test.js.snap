// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MerchantForm renders component by default 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="firstname"
      >
        First name
      </label>
      <input
        className="sc-AxjAm iCpIIE"
        id="firstname"
        name="firstname"
        placeholder=""
        type="text"
      />
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="lastname"
      >
        Last name
      </label>
      <input
        className="sc-AxjAm iCpIIE"
        id="lastname"
        name="lastname"
        placeholder=""
        type="text"
      />
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="email"
      >
        Email
      </label>
      <input
        className="sc-AxjAm iCpIIE"
        id="email"
        name="email"
        placeholder=""
        type="text"
      />
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="phone"
      >
        Phone number
      </label>
      <input
        className="sc-AxjAm iCpIIE"
        id="phone"
        name="phone"
        placeholder=""
        type="tel"
      />
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="hasPremium"
      >
        <div
          className="sc-Axmtr hMdFKE"
        >
          <input
            className="sc-AxhUy buggvC"
            id="hasPremium"
            name="hasPremium"
            type="checkbox"
          />
          <div
            checked={false}
            className="sc-fzoLsD kGKtmS"
          >
            <svg
              className="sc-AxmLO dXBjKZ"
              viewBox="0 0 24 24"
            >
              <polyline
                points="20 6 9 17 4 12"
              />
            </svg>
          </div>
          <span
            className="sc-fzozJi caPSEc"
          >
            Premium
          </span>
        </div>
      </label>
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw sc-AxgMl efeNvD"
        htmlFor="avatar"
      >
        <input
          accept="image/*"
          className="sc-AxhUy buggvC"
          id="avatar"
          name="avatar"
          type="file"
        />
        <img
          alt="Upload a file"
          className="sc-AxheI fmtaLC"
          src="upload.svg"
        />
        Upload a new avatar
      </label>
    </div>
    <div
      className="sc-fzplWN ksehmO"
    >
      
    </div>
    <input
      className="sc-fzpans OoBJa"
      onClick={[Function]}
      type="submit"
      value="Submit"
    />
  </form>
  <a
    href="/"
    onClick={[Function]}
  >
    Cancel
  </a>
</div>
`;

exports[`MerchantForm renders component with defined merchant 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="firstname"
      >
        First name
      </label>
      <input
        className="sc-AxjAm iCpIIE"
        id="firstname"
        name="firstname"
        placeholder=""
        type="text"
      />
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="lastname"
      >
        Last name
      </label>
      <input
        className="sc-AxjAm iCpIIE"
        id="lastname"
        name="lastname"
        placeholder=""
        type="text"
      />
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="email"
      >
        Email
      </label>
      <input
        className="sc-AxjAm iCpIIE"
        id="email"
        name="email"
        placeholder=""
        type="text"
      />
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="phone"
      >
        Phone number
      </label>
      <input
        className="sc-AxjAm iCpIIE"
        id="phone"
        name="phone"
        placeholder=""
        type="tel"
      />
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw gfimZx"
        htmlFor="hasPremium"
      >
        <div
          className="sc-Axmtr hMdFKE"
        >
          <input
            className="sc-AxhUy buggvC"
            id="hasPremium"
            name="hasPremium"
            type="checkbox"
          />
          <div
            checked={true}
            className="sc-fzoLsD wZUoW"
          >
            <svg
              className="sc-AxmLO dXBjKZ"
              viewBox="0 0 24 24"
            >
              <polyline
                points="20 6 9 17 4 12"
              />
            </svg>
          </div>
          <span
            className="sc-fzozJi caPSEc"
          >
            Premium
          </span>
        </div>
      </label>
    </div>
    <div
      className="sc-AxhCb gssqqQ"
    >
      <label
        className="sc-AxiKw sc-AxgMl efeNvD"
        htmlFor="avatar"
      >
        <input
          accept="image/*"
          className="sc-AxhUy buggvC"
          id="avatar"
          name="avatar"
          type="file"
        />
        <img
          alt="Upload a file"
          className="sc-AxheI fmtaLC"
          src="upload.svg"
        />
        Upload a new avatar
      </label>
    </div>
    <div
      className="sc-fzplWN ksehmO"
    >
      
      <img
        alt="Avatar preview"
        className="sc-fznyAO lhbMCo"
        src="avatar/url"
      />
    </div>
    <input
      className="sc-fzpans OoBJa"
      onClick={[Function]}
      type="submit"
      value="Submit"
    />
  </form>
  <a
    href="/"
    onClick={[Function]}
  >
    Cancel
  </a>
</div>
`;
